x/ibc-rate-limit/contracts/rate-limiter/src/packet.rs

```diff
@@ -30,35 +36,128 @@ pub struct Packet {
+    Eq,
+    ::prost::Message,
+    serde::Serialize,
+    serde::Deserialize,
+    schemars::JsonSchema,
+    CosmwasmExt,
+)]
+#[proto_message(type_url = "/cosmos.bank.v1beta1.QuerySupplyOf")]
+pub struct QuerySupplyOfResponse {
+    #[prost(message, optional, tag = "1")]
+    pub amount: ::core::option::Option<osmosis_std::types::cosmos::base::v1beta1::Coin>,
+}
+// End of SupplyOf query message definition
+
+use std::str::FromStr; // Needed to parse the coin's String as Uint256
+
 impl Packet {
-    pub fn channel_value(&self, _deps: Deps) -> u128 {
-        // let balance = deps.querier.query_all_balances("address", self.data.denom);
-        // deps.querier.sup
-        return 125000000000011250 * 2;
+    pub fn mock(channel_id: String, denom: String, funds: Uint256) -> Packet {
+        Packet {
+            sequence: 0,
+            source_port: "transfer".to_string(),
+            source_channel: channel_id.clone(),
+            destination_port: "transfer".to_string(),
+            destination_channel: channel_id,
+            data: crate::packet::FungibleTokenData {
+                denom,
+                amount: funds,
+                sender: Addr::unchecked("sender"),
+                receiver: Addr::unchecked("receiver"),
+            },
+            timeout_height: crate::packet::Height {
+                revision_number: None,
+                revision_height: None,
+            },
+            timeout_timestamp: None,
+        }
     }
 
-    pub fn get_funds(&self) -> u128 {
-        return self.data.amount;
+    pub fn channel_value(&self, deps: Deps) -> Result<Uint256, StdError> {
+        let res = QuerySupplyOfRequest {
+            denom: self.local_denom(),
+        }
+        .query(&deps.querier)?;
+        Uint256::from_str(&res.amount.unwrap_or_default().amount)
     }
 
-    fn local_channel(&self) -> String {
-        // Pick the appropriate channel depending on whether this is a send or a recv
-        return self.destination_channel.clone();
+    pub fn get_funds(&self) -> Uint256 {
+        self.data.amount
     }
 
-    fn local_demom(&self) -> String {
-        // This should actually convert the denom from the packet to the osmosis denom, but for now, just returning this
-        return self.data.denom.clone();
+    fn local_channel(&self, direction: &FlowType) -> String {
+        // Pick the appropriate channel depending on whether this is a send or a recv
+        match direction {
+            FlowType::In => self.destination_channel.clone(),
+            FlowType::Out => self.source_channel.clone(),
+        }
     }
 
-    pub fn path_data(&self) -> (String, String) {
-        let denom = self.local_demom();
-        let channel = if denom.starts_with("ibc/") {
-            self.local_channel()
-        } else {
-            "any".to_string() // native tokens are rate limited globally
-        };
+    fn local_denom(&self) -> String {
+        if !self.data.denom.starts_with("transfer/") {
```

Comment:
oh what, its transfer? I always thought it was `ibc/`

Reply:
`ibc/aaa...` is the name of the local token created by the transfer module, but the packet contains `port/channel/denom` (i.e.: `transfer/channel-32/osmo`). When receiving, the counterparty just sends their local denom, and the receiver hashes the token (after the middleware has handled it). When sending, we send `ibc/aaa...` but  the transfer module "unhashes" it before passing it down to the middleware. So the middleware always sees the denom trace

Reply:
ahh, I see. TIL! Can we add a comment somewhere to spec or go source code?